### PR TITLE
Initial new profiling work

### PR DESCRIPTION
**Pull request by blt**
This commit introduces a new performance test `file_to_blackhole`. What we have
here is not ready for prime time yet, but the basic stubs of a repeatable
process are in place. We have indications that users are struggling with file
source capped topologies keeping up, which is why I've dropped into blackhole
versus into an http sink. As much as possible this work focuses on file source
alone.

Throughput falls off pretty rapidly as line sizes increase, looks like. See
`notes.md` in this commit for details.

Signed-off-by: Brian L. Troutwine <brian@troutwine.us>

### CODE REVIEW

File: cases/performance/file_to_blackhole/notes.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+| 10 | 2.5 Mb/sec | 100 Kb | 25 Mb/sec | 25 Mb/sec | 25.7 Mb/sec | 0 Mb/sec |
+| 10 | 2.5 Mb/sec | 105 Kb | 25 Mb/sec | 23.9 Mb/sec | 24.6 Mb/sec | 1.1 Mb/sec |
+| 10 | 2.5 Mb/sec | 110 Kb | 25 Mb/sec | 22 Mb/sec | 22.06 Mb/sec | 3 Mb/sec |
+| 10 | 2.5 Mb/sec | 125 Kb | 25 Mb/sec | 17 Mb/sec | 17.04 Mb/sec | 9 Mb/sec |
+| 10 | 2.5 Mb/sec | 150 Kb | 25 Mb/sec | 12.3 Mb/sec | 12.6 Mb/sec | 12.7 Mb/sec |
+| 10 | 2.5 Mb/sec | 200 Kb | 25 Mb/sec | 6 Mb/sec | 6.15 Mb/sec | 19 Mb/sec |
+| 10 | 2.5 Mb/sec | 300 Kb | 25 Mb/sec | 3 Mb/sec | 3.08 Mb/sec | 22 Mb/sec |
+| 10 | 2.5 Mb/sec | 400 Kb | 25 Mb/sec | 1.53 Mb/sec | 1.57 Mb/sec | 23.47 Mb/sec |
+| 10 | 2.5 Mb/sec | 500 Kb | 25 Mb/sec | 1.05 Mb/sec | 1.07 Mb/sec | 23.95 Mb/sec |
+| 10 | 2.5 Mb/sec | 900 Kb | 25 Mb/sec | 231 Kb/sec | 237 Kb/sec | 24.77 Mb/sec |
```
Comment (blt) [inline cases/performance/file_to_blackhole/notes.md:70]: @lukesteensen @binarylogic this is probably the most relevant chunk for our ongoing conversation.

**Comment by blt** [comment]
@tsantero this is also potentially relevant to what you folks were experiencing.

**Comment by blt** [comment]
Relevant flamegraphs: https://gist.github.com/blt/87b90377ae8d8c6e8a08e28c6c437037

The size in the file names refers to the line size that `file_gen` produces.

**Comment by blt** [comment]
The diff doesn't mean much. The two flamegraphs for each line size are pretty radically different, which in itself is interesting.

**Comment by binarylogic** [comment]
That was fast, very nice. @lukesteensen has a much better understanding of the file source performance characteristics. I'm curious what he thinks.

**Comment by blt** [comment]
Worth calling out that this is not necessarily THE problem we’ve been working on, but it is a problem.

**Comment by tsantero** [comment]
@blt thanks for tagging me here. I'll be able to spend some time with this tonight and report back anything interesting.

**Comment by blt** [comment]
@tsantero rad, I appreciate it. If you pull 0.12 and flip debug logs on you'll get utilization numbers -- https://github.com/timberio/vector/pull/6634 -- that'll show us which part of your topology is consuming the most time.

File: cases/performance/file_to_blackhole/vector.toml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+data_dir = "data"
+
+[sources.internal]
+type = "internal_metrics"
+
+[sources.file]
+type = "file"
+max_line_bytes = 102400
```
Comment (lukesteensen) [inline cases/performance/file_to_blackhole/vector.toml:8]: Unless I'm missing something, I think this should explain why we see a drop-off right at 100kb line size 😄 . If we add another zero it should align better with the generated data.

That being said, I'm not terribly worried about extremely large individual lines. My assumption is that 100kb is a reasonable upper limit for "normal" use cases.

**Review by lukesteensen — COMMENTED** [review]
This looks great! Really happy to see this progressing.

File: cases/performance/file_to_blackhole/vector.toml (hunk 2 above)
Comment (blt) [reply cases/performance/file_to_blackhole/vector.toml:8]: D'oh, good eye. I thought I had that set to 1MB.

Comment (blt) [reply cases/performance/file_to_blackhole/vector.toml:8]: Well, that's good, correcting the line size actually makes the results I'm seeing more explicable. I'll add a section to the notes tomorrow. 

I can for sure produce a situation where vector falls behind and never catches up and this happens cleanly _without_ my `max_line_bytes` bug muddying the waters. If I have `file_gen` 10 files at 500Kb line bytes and adjust `max_line_bytes` to _actually_ be 1MB the effect still exists, except now the throughput from file source is 5MB where I'd expect 25MB. Likewise if `file_gen` produces 255 files at 100Kb line bytes the throughput from the file source is ~430Mb/s where `file_gen` is producing at ~630Mb/s.

Comment (blt) [reply cases/performance/file_to_blackhole/vector.toml:8]: If I adjust `file_gen` to produce 255 10Kb line files the numbers look like 600Mb/s from `file_gen` and 400Mb/s from the file source. Excellent result. I'll produce a clearer table tomorrow, but these numbers are promising.

File: cases/performance/file_to_blackhole/notes.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+One thing stood out in the tests: if I allowed vector to develop a backlog and
+then stopped `file_gen` the throughput would gradually catch up, before
+inevitably falling to zero when all the bytes were processed. There are two
+dimensions that cause the file source to fall behind, based on the above table:
+total number of files and byte throughput of files. 510 files of relatively
+meager throughput -- 48 Mb/sec -- forced file source to fall behind by 8 Mb/sec,
+a relatively small number of files -- 10 -- producing at 50Mb/sec caused the
+file source to fall behind by 254 Mb/sec. The progenitor code for `file-source`
+was concerned with fairness and not throughput, which I believe shows today in
+the data.
```
Comment (blt) [inline cases/performance/file_to_blackhole/notes.md:144]: @lukesteensen new results with the config bug corrected. Thanks again for catching that.

**Review by tobz — APPROVED** [review]
Nothing egregious jumps out here.

As discussed in chat, we know this is a first pass, originally designed for manual benchmarking, so we can continue to refine, standardize, etc.

**Pull request merged**